plugin loading path order

Hey Feri,

I have just been hit by an ordering issue when doing testing.

Basically the system had an old set of plugins installed in /usr/lib.... and it turns out that while I was executing some make check with new plugins, the checks were loading the system installed knet plugins instead of the newly built one.

this needs to be fixed or we need to have knet Buildconflict knet in the event we need to change internal API/ABI :-)

Hey Feri, finally back looking at this one.

Applied the following patch in 3 steps (one change at a time):

[fabbione@lilith kronosnet]$ git diff
diff --git a/libknet/Makefile.am b/libknet/Makefile.am
index 225d2c0..30fb134 100644
--- a/libknet/Makefile.am
+++ b/libknet/Makefile.am
@@ -85,6 +85,7 @@ EXTRA_libknet_la_DEPENDENCIES = $(SYMFILE)
 
 libknet_la_LDFLAGS     = -Wl,--version-script=$(srcdir)/$(SYMFILE) \
                          --export-dynamic \
+                         -Wl,--enable-new-dtags \
                          -Wl,-rpath=$(pkglibdir) \
                          -version-number $(libversion)
 
diff --git a/libknet/tests/Makefile.am b/libknet/tests/Makefile.am
index 13b6aec..28684ec 100644
--- a/libknet/tests/Makefile.am
+++ b/libknet/tests/Makefile.am
@@ -17,6 +17,7 @@ EXTRA_DIST            = \
 
 AM_CPPFLAGS            = -I$(top_srcdir)/libknet
 AM_CFLAGS              = $(PTHREAD_CFLAGS)
+AM_LDFLAGS             = -no-install -Wl,--enable-new-dtags
 LIBS                   = $(top_builddir)/libknet/libknet.la \
                          $(PTHREAD_LIBS) $(dl_LIBS)

as soon as I add the new-dtags to libknet/Makefile.am, no modules installed on the system, I am unable to run the test suite anylonger. It simply fails to find the plugins. Clearly neither RUNPATH or RPATH are being overwritten at this point. Adding stuff to tests/Makefile.am doesn´t change anything for this specific issue (except RPATH / RUNPATH are clearly visibile within objdump -p).

[root@kronosnet-node1-br0 kronosnet]# ./libtool --mode=execute strace -f -o /tmp/knet.strace libknet/tests/api_knet_send_crypto_test 
Using port 8243
Test knet_send with nss and valid data
knet_handle_crypto failed with correct config: Success
[knet]: [ERROR] common: unable to dlopen crypto_nss.so: crypto_nss.so: cannot open shared object file: No such file or directory
[knet]: [ERROR] crypto: Unable to load nss lib

[root@kronosnet-node1-br0 tests]# ./api_knet_send_crypto_test 
Using port 8254
Test knet_send with nss and valid data
knet_handle_crypto failed with correct config: Success
[knet]: [ERROR] common: unable to dlopen crypto_nss.so: crypto_nss.so: cannot open shared object file: No such file or directory
[knet]: [ERROR] crypto: Unable to load nss lib

[root@kronosnet-node1-br0 tests]# grep crypto_nss /tmp/knet.strace 
8243  open("/usr/lib64/kronosnet/tls/x86_64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/kronosnet/tls/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/kronosnet/x86_64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/kronosnet/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/lib64/tls/x86_64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/lib64/tls/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/lib64/x86_64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/lib64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/tls/x86_64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/tls/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/x86_64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  open("/usr/lib64/crypto_nss.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
8243  write(4, "unable to dlopen crypto_nss.so: "..., 256) = 256
8243  read(3, "unable to dlopen crypto_nss.so: "..., 256) = 256

[root@kronosnet-node1-br0 tests]# grep -i rpath /tmp/knet.strace 
[root@kronosnet-node1-br0 tests]# grep -i path /tmp/knet.strace 
[root@kronosnet-node1-br0 tests]#

--- a/configure.ac
+++ b/configure.ac
@@ -17,6 +17,12 @@ AC_INIT([kronosnet],
 	[devel@lists.kronosnet.org])
 AC_USE_SYSTEM_EXTENSIONS
 AM_INIT_AUTOMAKE([1.13 dist-bzip2 dist-xz color-tests -Wno-portability subdir-objects])
+
+# Switch to RUNPATH by default and disable RPATH
+# it is necessary to have this done very early in ./configure
+# before libtool does linker detection.
+# see also: https://github.com/kronosnet/kronosnet/issues/107
+LDFLAGS="$LDFLAGS -Wl,--enable-new-dtags"
 LT_PREREQ([2.2.6])
 LT_INIT
 

